### PR TITLE
#165412909 user signup

### PR DESCRIPTION
**Pull request by sengayire**
#### What does this PR do?
- Allow a user to register and get JWT token on a successful registration
-  User login using JWT strategy 
#### Description of Task to be completed?
Create the following endpoints:
POST /API/V1/users/signup
GET /API/V1/users/login
#### How should this be manually tested?
clone the repo locally and user postman to test endpoints

#### Any background context you want to provide?
- test endpoints with Mocha and Chai
#### What are the relevant pivotal tracker stories?
165412909

### CODE REVIEW

File: src/controllers/AuthPassportController.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      return {
+        code: status.CREATED,
+        user: newUser,
+        token: helper.tokenGenerator({ id: newUser.id, role: newUser.role })
+      };
+    }
+
+    return {
+      code: newUser.errors ? status.SERVER_ERROR : status.BAD_REQUEST,
+      error: 'account not created'
```
Comment (e-liyai) [inline src/controllers/AuthPassportController.js]: create `const` holders for constant values

File: src/controllers/AuthPassportController.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    const findUser = user.id && user.provider
+      ? await User.findOne({
+        accountProvider: user.provider,
+        accountProviderUserId: user.id
+      })
+      : {};
+
+    if (findUser.errors) {
+      return res.status(status.SERVER_ERROR).json({
+        error: 'sorry, something went wrong'
```
Comment (e-liyai) [inline src/controllers/AuthPassportController.js]: same here

File: src/controllers/AuthLocalController.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+   */
+  static async login(req, res) {
+    try {
+      const { email, password } = req.body;
+      const checkUser = await User.findOne({ email });
+      if (Object.keys(checkUser).length > 0) {
+        const comparePassword = helper.password.compare(password, checkUser.password);
+        if (!comparePassword) {
+          return res.status(status.UNAUTHORIZED).send({
+            message: 'The credentials you provided is incorrect'
```
Comment (e-liyai) [inline src/controllers/AuthLocalController.js]: create `const` holders for constant values

File: src/helpers/factory/article.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { Factory } from 'rosie';
+
+export default Factory.define('article')
```
Comment (e-liyai) [inline src/helpers/factory/article.js]: use `chance` to generate content

File: src/controllers/AuthPassportController.js (hunk 1 above)
Comment (sengayire) [reply src/controllers/AuthPassportController.js]: please, what do you mean by const holders?

File: src/helpers/factory/article.js (hunk 4 above)
Comment (sengayire) [reply src/helpers/factory/article.js]: I am not the one who worked on  the article so, I think I should not change it to my commit but I can still do it if there is no problem, thank you

File: src/controllers/AuthPassportController.js (hunk 1 above)
Comment (rwajon) [reply src/controllers/AuthPassportController.js]: On my understanding those error messages should be defined as constants then they can be reused any where

Comment (sengayire) [reply src/controllers/AuthPassportController.js]: Ok good, I am working on it, thank you

**Comment by luc-tuyishime** [comment]
I can see files are conflicting with develop, so try to rebase first and continue working on your local branch.

File: src/helpers/factory/article.js (hunk 4 above)
Comment (sengayire) [reply src/helpers/factory/article.js]: This has been implemented

File: src/controllers/AuthLocalController.js (hunk 3 above)
Comment (sengayire) [reply src/controllers/AuthLocalController.js]: this has been implemented

File: src/tests/helpers/passportLocal.test.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import db from '../../models';
+import passport from '../../config/passportLocalConfig';
+
+chai.should();
+
+chai.use(chaiHttp);
+let user = {};
+const newUser = {
```
Comment (e-liyai) [inline src/tests/helpers/passportLocal.test.js]: use factories to generate data

File: src/tests/controllers/users.test.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+import app from '../../app';
+import status from '../../config/status';
+import db from '../../models';
+
+chai.should();
+chai.use(chaiHttp);
+
+// user test
+describe('user tests', () => {
+  const user = {
```
Comment (e-liyai) [inline src/tests/controllers/users.test.js]: use factories to generate data

File: src/tests/controllers/users.test.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+chai.should();
+chai.use(chaiHttp);
+
+// user test
+describe('user tests', () => {
+  const user = factory.user.build();
+
+  // unvalid user
+  const invalidUser = {
```
Comment (e-liyai) [inline src/tests/controllers/users.test.js]: @sengayire what about this one?

File: src/tests/middlewares/validateNewUser.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (e-liyai) [inline src/tests/middlewares/validateNewUser.js]: this test file should be named according to the conventions we agreed on

**Comment by e-liyai** [comment]
@sengayire just realized your test files don't follow the conventions we agreed on

**Pull request merged**